File: patterns/factory/UserAndAssessmentFactory.js
/**
 * DESIGN PATTERN IMPLEMENTATION: Factory Method Pattern
 * 
 * Pattern: Factory Method (Creational)
 * Purpose: Defines an interface for creating objects, but lets subclasses or factory methods
 *          decide which class to instantiate. Solves the issue of direct object instantiation 
 *          for different user roles and educational assessment types.
 */

// Base User Class
class BaseUser {
    constructor(data) {
        this.id = data.id;
        this.name = data.name;
        this.email = data.email;
        this.role = data.role;
    }

    getPermissions() {
        return [];
    }
}

class TutorUser extends BaseUser {
    constructor(data) {
        super(data);
        this.specialization = data.specialization || 'General Subject Tutor';
        this.bio = data.bio || '';
    }

    getPermissions() {
        return ['CREATE_BATCH', 'SCHEDULE_SESSION', 'UPLOAD_MATERIAL', 'CREATE_ASSIGNMENT', 'GRADE_SUBMISSION', 'CREATE_QUIZ', 'MARK_ATTENDANCE'];
    }
}

class StudentUser extends BaseUser {
    constructor(data) {
        super(data);
        this.institution = data.institution || 'University';
        this.academicLevel = data.academic_level || 'Undergraduate';
    }

    getPermissions() {
        return ['VIEW_SCHEDULE', 'SUBMIT_ASSIGNMENT', 'ATTEMPT_QUIZ', 'VIEW_GRADES', 'REQUEST_SCHEDULE_CHANGE', 'CHAT_TUTOR'];
    }
}

class ParentUser extends BaseUser {
    constructor(data) {
        super(data);
        this.occupation = data.occupation || 'Guardian';
    }

    getPermissions() {
        return ['VIEW_CHILD_SCHEDULE', 'VIEW_CHILD_PROGRESS', 'VIEW_CHILD_ATTENDANCE', 'CHAT_TUTOR'];
    }
}

// User Factory
class UserFactory {
    static createUser(data) {
        switch (data.role) {
            case 'TUTOR':
                return new TutorUser(data);
            case 'STUDENT':
                return new StudentUser(data);
            case 'PARENT':
                return new ParentUser(data);
            default:
                throw new Error(`Unknown user role: ${data.role}`);
        }
    }
}

// Assessment Classes & Factory
class QuizAssessment {
    constructor(data) {
        this.id = data.id;
        this.title = data.title;
        this.timeLimitMins = data.time_limit_mins || 30;
        this.totalMarks = data.total_marks || 100;
        this.type = 'QUIZ';
    }

    calculateGrade(submittedAnswers, answerKey) {
        // Automatic MCQ Grading logic
        let score = 0;
        for (const [qId, ans] of Object.entries(submittedAnswers)) {
            if (answerKey[qId] && answerKey[qId].correct === ans) {
                score += answerKey[qId].marks || 10;
            }
        }
        return score;
    }
}

class AssignmentAssessment {
    constructor(data) {
        this.id = data.id;
        this.title = data.title;
        this.deadline = data.deadline;
        this.totalMarks = data.total_marks || 100;
        this.type = 'ASSIGNMENT';
    }

    calculateGrade(manualScore, feedback) {
        return {
            score: Math.min(manualScore, this.totalMarks),
            feedback: feedback || 'Graded by tutor.'
        };
    }
}

class AssessmentFactory {
    static createAssessment(type, data) {
        if (type === 'QUIZ') {
            return new QuizAssessment(data);
        } else if (type === 'ASSIGNMENT') {
            return new AssignmentAssessment(data);
        }
        throw new Error(`Unsupported assessment type: ${type}`);
    }
}

module.exports = {
    UserFactory,
    AssessmentFactory,
    TutorUser,
    StudentUser,
    ParentUser,
    QuizAssessment,
    AssignmentAssessment
};

File: docs/java_patterns/UserAndAssessmentFactory.java
package com.tms.patterns.factory;

import java.util.*;

/**
 * DESIGN PATTERN IMPLEMENTATION: Factory Method Pattern (Java)
 * 
 * Pattern: Factory Method (Creational)
 * Purpose: Defines an interface for creating objects, but lets subclasses or factory methods
 *          decide which class to instantiate for User types and Assessment types.
 */

abstract class User {
    protected int id;
    protected String name;
    protected String email;
    protected String role;

    public User(int id, String name, String email, String role) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.role = role;
    }

    public abstract List<String> getPermissions();
}

class TutorUser extends User {
    private String specialization;

    public TutorUser(int id, String name, String email, String specialization) {
        super(id, name, email, "TUTOR");
        this.specialization = specialization;
    }

    @Override
    public List<String> getPermissions() {
        return Arrays.asList("CREATE_BATCH", "SCHEDULE_SESSION", "UPLOAD_MATERIAL", "CREATE_ASSIGNMENT", "GRADE_SUBMISSION", "CREATE_QUIZ", "MARK_ATTENDANCE");
    }
}

class StudentUser extends User {
    private String institution;

    public StudentUser(int id, String name, String email, String institution) {
        super(id, name, email, "STUDENT");
        this.institution = institution;
    }

    @Override
    public List<String> getPermissions() {
        return Arrays.asList("VIEW_SCHEDULE", "SUBMIT_ASSIGNMENT", "ATTEMPT_QUIZ", "VIEW_GRADES", "REQUEST_SCHEDULE_CHANGE", "CHAT_TUTOR");
    }
}

class ParentUser extends User {
    private String occupation;

    public ParentUser(int id, String name, String email, String occupation) {
        super(id, name, email, "PARENT");
        this.occupation = occupation;
    }

    @Override
    public List<String> getPermissions() {
        return Arrays.asList("VIEW_CHILD_SCHEDULE", "VIEW_CHILD_PROGRESS", "VIEW_CHILD_ATTENDANCE", "CHAT_TUTOR");
    }
}

// User Factory
public class UserAndAssessmentFactory {
    public static User createUser(int id, String name, String email, String role, String extraDetail) {
        switch (role.toUpperCase()) {
            case "TUTOR":
                return new TutorUser(id, name, email, extraDetail);
            case "STUDENT":
                return new StudentUser(id, name, email, extraDetail);
            case "PARENT":
                return new ParentUser(id, name, email, extraDetail);
            default:
                throw new IllegalArgumentException("Unknown role: " + role);
        }
    }
}
